agent/consul/grpc_resolver.go

```diff
@@ -0,0 +1,105 @@
+package consul
+
+import (
+	"sync"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"google.golang.org/grpc/resolver"
+)
+
+const grpcResolverScheme = "consul"
+
+var resolverBuilder *ServerResolverBuilder
+
+func init() {
+	resolverBuilder = NewServerResolverBuilder()
+	resolver.Register(resolverBuilder)
+}
+
+// ServerResolverBuilder tracks the current server list and keeps any
+// ServerResolvers updated when changes occur.
+type ServerResolverBuilder struct {
+	servers   map[string]*metadata.Server
+	resolvers map[*ServerResolver]struct{}
+	lock      sync.Mutex
+}
+
+func NewServerResolverBuilder() *ServerResolverBuilder {
+	return &ServerResolverBuilder{
+		servers:   make(map[string]*metadata.Server),
+		resolvers: make(map[*ServerResolver]struct{}),
+	}
+}
+
+func (s *ServerResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOption) (resolver.Resolver, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// Make a new resolver and add it to the list of active ones.
```

Comment:
Why would there be multiple active ones and why can't they all just share? Is there one per server conn?

Reply:
Initially I had only the one resolver that got returned every time, but I got issues with parallel tests clobbering each other since the builder gets initialized at the package level. I think we may want to be able to handle multiple resolvers anyway here to be able to pass in a Target of `consul:///server.dc2` or something similar to handle cross DC requests, which I should include in this PR.